### PR TITLE
Upgrading `torchsparse` to version 2.1.0

### PR DESCRIPTION
**Pull request by kellman**
Fetched the upstream master branch from `torchsparse` and merged it into our fork, branch `zendar-main`. I upgraded the python and build dependencies in the `pyproject.toml` file. 

I was running into a compile issue, it was very non-specific. It was an our of memory issue when using `ninja`. I set the `MAX_JOBS=1` to limit the parallelism. This did the trick.

I added this to precommit workflow. I was not sure if this would be an issue on the CI machine. I can leave it out, but it will just be slower (but less prone to crash, unless it has another way to handle this issue).

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -23,4 +23,4 @@ dynamic = ["version"]
 version = {attr = "torchsparse.__version__"}
 
 [build-system]
-requires = ["torch==2.2.0+cu121", "tqdm", "numpy", "setuptools"]
+requires = ["torch==2.2.0+cu121", "tqdm", "numpy", "setuptools", "wheel", "rootpath", "attributedict", "typing-extensions"]
```
Comment (dranjan) [inline pyproject.toml]: Are all of these really build dependencies?

**Review by dranjan — APPROVED** [review]
The few changes here on top of the upstream seem acceptable to me, but it would be nice if we could simplify the Zendar-specific tooling somehow and bring our fork closer to vanilla torchsparse.

**Comment by dranjan** [comment]
BTW @kellman - I would recommend reverting your last two commits because they aren't strictly necessary to have the build succeed, and they're very likely to complicate pulling future updates. I wouldn't worry about the failing GitHub action. Even the upstream isn't passing pre-commit.

File: pyproject.toml (hunk 1 above)
Comment (kellman) [reply pyproject.toml]: I removed all but the additional `rootpath` dep.

File: setup.py [hunk 2]
```diff
@@ -4,38 +4,58 @@
...
+    CUDA_HOME,
+    BuildExtension,
+    CppExtension,
+    CUDAExtension,
+)
+
+# from torchsparse import __version__
 
-from torchsparse import __version__
+version_file = open("./torchsparse/version.py")
```
Comment (ishivvers) [inline setup.py:16]: Ok - it looks like this is the line where [they parse the `__version__` from `torchsparse/version.py`](https://github.com/ZendarInc/torchsparse/blob/mk/v2.1.0/torchsparse/version.py)

No idea why they're reading the python file as a text file and parsing it by hand - their `master-v2.2.0` branch does as you do here, but [the `master` branch (which hasn't been touched for years) does it a more sane way](https://github.com/ZendarInc/torchsparse/blob/1a10fda15098f3bf4fa2d01f8bee53e85762abcf/setup.py#L10).


-----------------------------------------------

Anyways, regardless of how the file-reading is done, can you make sure that the version declared in `torchsparse/version.py` is what's correctly getting allocated to the built wheel and saved into the filename?  That's how we'll be versioning `torchsparse` as we use it in `RadarProcessor`.

Comment (dranjan) [reply setup.py:16]: Are you sure you can import `torchsparse` before the package has been built? In general that's not a safe assumption.

Comment (ishivvers) [reply setup.py:16]: ah - great - yeah @dranjan maybe that's why they're parsing the `version.py` file as a text file :heavy_check_mark:

Comment (kellman) [reply setup.py:16]: okay I do not think this is an issue. The built package has the correct version.
```
iPython 3.11.7 (main, Dec  8 2023, 18:56:57) [GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torchsparse
>>> torchsparse.__version__
'2.1.0'
```

Comment (ishivvers) [reply setup.py:16]: Great - thanks for checking.

Then please just ensure you follow [the directions to upload the built wheel](https://github.com/ZendarInc/torchsparse?tab=readme-ov-file#rolling-out-updates-at-zendar) once this is merged, including the new version in the github release, and then the versioning should all work out!

**Review by ishivvers — APPROVED** [review]

**Pull request merged**